api/src/main/java/org/openmrs/validator/DrugOrderValidator.java

```diff
@@ -93,58 +97,64 @@ public void validate(Object obj, Errors errors) {
 			if (order.getDrug() == null || order.getDrug().getConcept() == null) {
 				ValidationUtils.rejectIfEmpty(errors, "concept", "error.null");
 			}
-			
+
 			if (order.getConcept() != null && order.getDrug() != null && order.getDrug().getConcept() != null
-			        && !order.getDrug().getConcept().equals(order.getConcept())) {
-				errors.rejectValue("drug", "error.general");
-				errors.rejectValue("concept", "error.concept");
-			}
+					&& !order.getDrug().getConcept().equals(order.getConcept())) {
+             errors.rejectValue("drug", "error.general");
+             errors.rejectValue("concept", "error.concept");
+
+			} 
 			if (order.getAction() != Order.Action.DISCONTINUE && order.getDosingType() != null) {
 				DosingInstructions dosingInstructions = order.getDosingInstructionsInstance();
 				dosingInstructions.validate(order, errors);
 			}
+			if (order.getConcept() == null && order.getDrugNonCoded() != null) {
+	             if (order.getDrugNonCoded().equals(Context.getOrderService().getNonCodedDrugConcept())) {
+	                   order.setConcept(Context.getOrderService().getNonCodedDrugConcept());
```

Comment:
HI @ibacher  @dkayiwa ,what could be your take on this in regards with the ticket.Steps i tried to accomplish include
1)Identifying a Non Coded Drug
2)Setting the concept basing on the non coded coded drug

Reply:
I'm a little hesitant to endorse changing the order in a validator, but I suppose it needs to be done. Note, however, that the condition wrapping this line will always be false, i.e., `order.getDrugNonCoded()` (which returns a String) will **never** be equal to `Context.getOrderService().getNonCodedDrugConcept()` (which returns a Concept).

Reply:
I agree with @ibacher on his reservation of updating the drug order here. I think the best place to update the concept would be in the associated setter for `DrugNonCode` in the `DrugOrder` class https://github.com/openmrs/openmrs-core/blob/9fcd6677f0735c047896b42488d06cd7db5b8432/api/src/main/java/org/openmrs/DrugOrder.java#L546. 

```
IF value to set is not empty
  IF drugOrder.concept is null
    SET drugOrder.concept to NonCodedDrugConcept
  END IF
ELSE
  IF drugOrder.concept is not null
    IF drugOrder.concept equals NonCodedDrugConcept 
      SET drugOrder.concept to null
    END IF
  END IF
END IF
```